kratos/python_scripts/auxiliar_methods_adaptative_remeshing.py

```diff
@@ -0,0 +1,161 @@
+        {
+            "list_model_parts_to_assign_conditions" : []
+        }
+        """)
+        root_model_part_name = computing_model_part.GetRootModelPart().Name
+        for name in list_model_parts:
+            name = name.replace(root_model_part_name + ".", "")
+            skin_detection_parameters["list_model_parts_to_assign_conditions"].Append(name)
+
+        if computing_model_part.ProcessInfo[KratosMultiphysics.DOMAIN_SIZE] == 2:
+            detect_skin = KratosMultiphysics.SkinDetectionProcess2D(computing_model_part, skin_detection_parameters)
+        else:
+            detect_skin = KratosMultiphysics.SkinDetectionProcess3D(computing_model_part, skin_detection_parameters)
+        detect_skin.Execute()
+    solver.SetEchoLevel(analysis.echo_level)
+
+def AdaptativeRemeshingRunSolutionLoop(analysis):
+    """This function executes the solution loop of the AnalysisStage for cases where remeshing may be considered
+
+        Keyword arguments:
+        analysis The AnalysisStage to be reinitialized
+    """
+
+    # If we remesh using a process
+    computing_model_part = analysis._GetSolver().GetComputingModelPart()
+    root_model_part = computing_model_part.GetRootModelPart()
+
+    while analysis.KeepAdvancingSolutionLoop():
+        analysis.time = analysis._GetSolver().AdvanceInTime(analysis.time)
+        # We reinitialize if remeshed previously
+        if root_model_part.Is(KratosMultiphysics.MODIFIED):
+            ReInitializeSolver(analysis)
+        analysis.InitializeSolutionStep()
+        # We reinitialize if remeshed on the InitializeSolutionStep
+        if root_model_part.Is(KratosMultiphysics.MODIFIED):
+            ReInitializeSolver(analysis)
+            analysis.InitializeSolutionStep()
+        analysis._GetSolver().Predict()
+        analysis._GetSolver().SolveSolutionStep()
+        analysis.FinalizeSolutionStep()
+        analysis.OutputSolutionStep()
+
+def SPRAdaptativeRemeshingRunSolutionLoop(analysis, with_contact = False):
+    """This function executes the solution loop of the AnalysisStage for cases where remeshing may be considered with SPR convergence criteria
+
+        Keyword arguments:
+        analysis The AnalysisStage to be reinitialized
+    """
+
+    # Remeshing adaptively
+    solver = analysis._GetSolver()
+    computing_model_part = solver.GetComputingModelPart()
+    root_model_part = computing_model_part.GetRootModelPart()
+    metric_process = solver.get_metric_process()
+    remeshing_process = solver.get_remeshing_process()
+    convergence_criteria = solver.get_convergence_criterion()
+    builder_and_solver = solver.get_builder_and_solver()
+    mechanical_solution_strategy = solver.get_mechanical_solution_strategy()
+
+    while analysis.KeepAdvancingSolutionLoop():
+        analysis.time = solver.AdvanceInTime(analysis.time)
+        non_linear_iteration = 1
+        while non_linear_iteration <= analysis.non_linear_iterations:
```

Comment:
which non-linear iteration is this ? for the Newton-raphson  ?

Reply:
Yes